src/Umbraco.Cms.Api.Management/Services/BackOfficeExternalLoginService.cs

```diff
@@ -146,6 +148,52 @@ public async Task<Attempt<ExternalLoginOperationStatus>> UnLinkLoginAsync(Claims
         return Attempt.SucceedWithStatus(ExternalLoginOperationStatus.Success, Enumerable.Empty<IdentityError>());
     }
 
+    public async Task<Attempt<Guid?, ExternalLoginOperationStatus>> GenerateLoginProviderSecretAsync(ClaimsPrincipal claimsPrincipal, string loginProvider)
+    {
+        if (claimsPrincipal.Identity is null)
+        {
+            return Attempt.FailWithStatus<Guid?, ExternalLoginOperationStatus>(ExternalLoginOperationStatus.IdentityNotFound, null);
+        }
+
+        IEnumerable<BackOfficeExternaLoginProviderScheme> configuredLoginProviders = await _backOfficeExternalLoginProviders.GetBackOfficeProvidersAsync();
+        if (configuredLoginProviders.Any(provider => provider.ExternalLoginProvider.AuthenticationType.Equals(loginProvider))
+            is false)
+        {
+            return Attempt.FailWithStatus<Guid?, ExternalLoginOperationStatus>(ExternalLoginOperationStatus.AuthenticationSchemeNotFound, null);
+        }
+
+        var userId = claimsPrincipal.Identity.GetUserId();
+        if (userId is null)
+        {
+            return Attempt.FailWithStatus<Guid?, ExternalLoginOperationStatus>(ExternalLoginOperationStatus.IdentityNotFound, null);
+        }
+
+        var secret = Guid.NewGuid();
+        _memoryCache.Set(secret, new LoginProviderUserLink { ClaimsPrincipal = claimsPrincipal, LoginProvider = loginProvider }, new MemoryCacheEntryOptions { AbsoluteExpirationRelativeToNow = TimeSpan.FromSeconds(30) });
```

Comment:
Is it safe to cache the actual claims principal here? Is there any chance of state/scope mismatch on subsequent requests? An alternative would be to cache the user ID and use that to re-generate the matching claims principal in `ClaimsPrincipleFromLoginProviderLinkKey`

Reply:
Since it is a very short lived cache, i felt it would be ok to just cache it as the goal is to split up the same request into 2 bits.
But it do share your concern, can change it if you deem it necessary.